python/social_bot/tasks.py

```diff
@@ -968,21 +975,21 @@ def run(self):
             r_contact = self._agent.get_contacts('finger_cnta_r',
                                                  self._object_collision_name)
             # check distance and contacts
-            dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
-            palm_dist = np.linalg.norm(
-                np.array(obj_pos) - np.array(finger_pos))
-            gripping_feature_num = 1.0 * l_contact + 1.0 * r_contact + 10.0 * min(
-                obj_pos[2], 0.1)  # encourge to lift the object
-            gripping = (gripping_feature_num >= 2.5)
+            obj_dist = np.linalg.norm(np.array(obj_pos) - goal_pos)
+            palm_dist = np.linalg.norm(np.array(obj_pos) - np.array(finger_pos))
+            obj_lifted = obj_height / self._obj_init_height - 1.0
+            gripping_feature = 0.25 * l_contact + 0.25 * r_contact + min(
+                obj_lifted, 1.0)  # encourge to lift the object by obj_height
+            gripping = (gripping_feature >= 0.999999)
             # yield rewards
-            if gripping and dist < self._success_distance_thresh:
+            if gripping and obj_dist - self._obj_init_height < self._success_distance_thresh:  # minus an offset of object height
                 logging.debug("object has been successfuly placed")
                 reward = 100.0 if self._reward_shaping else 1.0
                 agent_sentence = yield TeacherAction(
                     reward=reward, sentence="well done", done=True)
             else:
-                shaped_reward = (3.5 + obj_pos[2] - min(dist, 1.0)) if gripping else (
-                    gripping_feature_num - palm_dist)
+                shaped_reward = 2.0 - min(obj_dist, 1.0) if gripping else (
```

Comment:
if obj_dist is more than 1, object is faraway, then the agent will still get shaped_reward of 1?  Will this work?

Reply:
Yes, this is on purpose. If not griping, the agent can get a maximal reward of 1 (gripping_feature). State of gripping should be at least better than not griping.

Reply:
I see, it encourages gripping.  However, it doesn't help guide the agent get closer to the target (after gripping object)?  Since we are doing reward shaping already, why not encourage agent to grip and get close to goal by using a strictly increasing reward as it gets closer to goal?

Reply:
It did help guide the agent get closer to the target after gripping object, if obj_dist is less than 1 meters. If obj_dist is more higher, it is still much easier than _success_distance_thresh.

Reply:
Using strictly increasing reward as getting closer to goal is another option, but it is a little bit complicated: 

The distance getting closer to goal is something like (self._prev_dist - dist). 这个数值很小,一条成功的trajectory所有的pre_dist - dist 总的加起来, 最大也就是random range,即0.5, 而把cube倾斜一点然后抬起来就很容易接近1.0 per step. 所以也有同样的问题, 导致agent不去抓起来而只是按在地上摩擦. 所以我们也要加上1.0的base: 1.0 + max(self._prev_dist - dist, 0).

考虑一条成功的trajectory所有的pre_dist - dist 总的加起来最大也就只有random range,在这里默认参数是0.5, 而为griping设计的reward可以达到1.0*maxsteps steps, 会到200. 信噪比太低了, 200/0.5. 所以还需要scale一下, 变成 1.0 + 400.0 * max(self._prev_dist - dist, 0).

最后的success也会有同样的问题,如果还用1.0, 这个1.0对agent就没有什么吸引力了,因为抓着object都比1.0大,所以reward shaping的时候也改成了100. 

但是100也可能会有问题, 因为我们这个任务是手动设计的结束状态(靠maxsteps结束而不是gym的max_episode_steps),agent要maximize discounted future return, 如果future的reward discount下来相比迅速success然后结束时的100不可以被忽略,最佳策略变成了抓起来object,然后接近目标,然后等待,等到最后一个step的时候去success. (实际训练的时候确实也发现了这个现象).

那好我们把100继续加大,加到10000, 最终第200个step的10000 discount下来加上future的每一步的reward discount合起来,没有立刻加上这不带discount的10000大,agent就会去迅速去success然后end了.

但是后期rewad幅度过大可能会带来一些数值问题,比如对policy gradient method的loss来说, scale reward等效于去sacle 学习率. SAC的话就会更凌乱一点, 相当于只去sacle return有关部分的学习率, entropy部分没有scale. 而且这个sacle是训练前期学习率小,后期学习率大...

所以只用 strictly increasing reward 容易导致后面的reward大的更厉害. 对比较多任务阶段的reward shaping, 就会像社会发展的通货膨胀一样,越来越大. 比如现在的PR2 gripper的reward shaping, 幅度变化非常大,两个griper touch 到object, 累计reward在个位数,然后抓起来的话累计reward要到几百. 我觉得这可能是导致某些算法训练比较慢或者后期变得不稳定的原因之一.(小的学习率速度慢,大的学习率后期不稳定.)